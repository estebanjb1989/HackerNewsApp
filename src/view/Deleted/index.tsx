import React, { useMemo } from 'react';
import {
  SafeAreaView,
  FlatList,
  StyleSheet,
  View,
} from 'react-native';
import ListItem from "@view/Articles/ListItem"
import Title from "@component/Title"
import _ from "lodash"
import { IArticle, IState } from '@interfaces';
import { useSelector } from 'react-redux';

const Separator = () => <View style={{ height: 10 }} />;

function Deleted() {  
  const [
    articles,
    deletedArticles,    
  ] = useSelector((state: IState) => [
    state.articles.articles,
    state.articles.deletedArticles,    
  ])

  const articlesMemo = useMemo(() => {
    return articles.filter((article: IArticle) =>
      deletedArticles?.some((da: IArticle) => da === article.id)
    )
  }, [articles, deletedArticles])


  return (
    <SafeAreaView>
      <Title text="Deleted Articles" />
      <FlatList
        style={styles.flatList}
        data={articlesMemo}
        keyExtractor={(item) => item.id}
        windowSize={10}
        renderItem={({ item }) => (
          <ListItem
            key={item.title}
            item={item}
            isFavorite={true}
          />
        )}
        ItemSeparatorComponent={Separator}
      />

    </SafeAreaView>
  );
}

const styles = StyleSheet.create({
  flatList: {
    backgroundColor: 'white'
  },
});

export default Deleted;
