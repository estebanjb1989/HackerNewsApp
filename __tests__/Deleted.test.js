import React from 'react';
import { render } from '@testing-library/react-native';
import { useSelector } from "react-redux"
import Deleted from "@view/Deleted"

const mockState = {
  articles: {
    articles: [{
      id: "1111111111",
      author: "author1",
      title: "article1",
      url: "www.google.com"
    }, {
      id: "2222222222",
      author: "author2",
      title: "article2",
      url: "www.google.com"
    }],
    favoriteArticles: [],
    deletedArticles: [
      "1111111111"
    ]
  }
}

jest.mock("react-redux", () => ({
  useSelector: jest.fn()
}))

jest.mock("@react-navigation/native", () => ({
  useNavigation: jest.fn()
}))

describe("Deleted Articles view", () => {
  beforeEach(() => {
    useSelector.mockImplementation((cb) => cb(mockState))
  })

  afterEach(() => {
    jest.clearAllMocks()
  })

  it('renders correctly', () => {
    const tree = render(<Deleted />);
    expect(tree).toMatchSnapshot();
  });

  it('should render article present in deleted list', async () => {
    const tree = render(<Deleted />);
    const authorText = await tree.queryByTestId("list-item_author")    
    expect(authorText.props?.children).toBe("author1")
  });

  it('should not render article not present in deleted list', async () => {
    const tree = render(<Deleted />);
    const authorTextNodes = await tree.queryAllByTestId("list-item_author")    
    for(const node of authorTextNodes) {
      expect(node.props?.children).not.toBe("author2")  
    }
  });
})

