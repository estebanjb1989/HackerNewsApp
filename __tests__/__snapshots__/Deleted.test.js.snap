// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Deleted Articles view renders correctly 1`] = `
<RCTSafeAreaView>
  <Text
    style={
      [
        {
          "fontSize": 20,
          "marginHorizontal": 8,
          "marginVertical": 8,
          "minWidth": 240,
        },
        undefined,
      ]
    }
  >
    Deleted Articles
  </Text>
  <RCTScrollView
    ItemSeparatorComponent={[Function]}
    data={
      [
        {
          "author": "author1",
          "id": "1111111111",
          "title": "article1",
          "url": "www.google.com",
        },
      ]
    }
    getItem={[Function]}
    getItemCount={[Function]}
    keyExtractor={[Function]}
    onContentSizeChange={[Function]}
    onLayout={[Function]}
    onMomentumScrollBegin={[Function]}
    onMomentumScrollEnd={[Function]}
    onScroll={[Function]}
    onScrollBeginDrag={[Function]}
    onScrollEndDrag={[Function]}
    removeClippedSubviews={false}
    renderItem={[Function]}
    scrollEventThrottle={0.0001}
    stickyHeaderIndices={[]}
    style={
      {
        "backgroundColor": "white",
      }
    }
    viewabilityConfigCallbackPairs={[]}
    windowSize={10}
  >
    <View>
      <View
        onFocusCapture={[Function]}
        onLayout={[Function]}
        style={null}
      >
        <View
          accessibilityState={
            {
              "busy": undefined,
              "checked": undefined,
              "disabled": undefined,
              "expanded": undefined,
              "selected": undefined,
            }
          }
          accessibilityValue={
            {
              "max": undefined,
              "min": undefined,
              "now": undefined,
              "text": undefined,
            }
          }
          accessible={true}
          collapsable={false}
          focusable={true}
          onClick={[Function]}
          onResponderGrant={[Function]}
          onResponderMove={[Function]}
          onResponderRelease={[Function]}
          onResponderTerminate={[Function]}
          onResponderTerminationRequest={[Function]}
          onStartShouldSetResponder={[Function]}
          style={
            {
              "opacity": 1,
            }
          }
        >
          <View
            style={
              {
                "alignItems": "center",
                "borderColor": "grey",
                "borderRadius": 6,
                "borderStyle": "solid",
                "borderWidth": 0.5,
                "flexDirection": "row",
                "gap": 12,
                "marginHorizontal": 8,
                "padding": 12,
              }
            }
          >
            <View
              style={
                {
                  "flex": 1,
                }
              }
            >
              <View
                style={
                  {
                    "flexDirection": "row",
                    "justifyContent": "space-between",
                  }
                }
              >
                <Text>
                  Invalid Date
                </Text>
                <Image
                  source={
                    {
                      "testUri": "../../../src/img/icons/favorite.png",
                    }
                  }
                  style={
                    {
                      "height": 32,
                      "width": 32,
                    }
                  }
                />
              </View>
              <Text>
                article1
              </Text>
              <Text
                testID="list-item_author"
              >
                author1
              </Text>
            </View>
          </View>
        </View>
      </View>
    </View>
  </RCTScrollView>
</RCTSafeAreaView>
`;
